request.go

```diff
@@ -0,0 +1,111 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tspclient
+
+import (
+	"crypto"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/notaryproject/tspclient-go/internal/hashutil"
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// MessageImprint contains the hash of the datum to be time-stamped.
+//
+//	MessageImprint ::= SEQUENCE {
+//	 hashAlgorithm   AlgorithmIdentifier,
+//	 hashedMessage   OCTET STRING }
+type MessageImprint struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	HashedMessage []byte
+}
+
+// Request is a time-stamping request.
+//
+//	TimeStampReq ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 messageImprint  MessageImprint,
+//	 reqPolicy       TSAPolicyID              OPTIONAL,
+//	 nonce           INTEGER                  OPTIONAL,
+//	 certReq         BOOLEAN                  DEFAULT FALSE,
+//	 extensions      [0] IMPLICIT Extensions  OPTIONAL }
+type Request struct {
```

Comment:
Should we add a method to validate correctness of Request? like version is 1 etc

Reply:
@priteshbandi If user chooses to use `NewRequest` or `NewRequestFromContent`, then the `Request` sent to TSA is guaranteed to align with RFC 3161. However, if caller chooses to use self-build TSA `Request`, and if it's an invalid request, `Timestamper.Timstamp(context.Context, *Request)` would fail, in other words, the check is done by the TSA.

Reply:
Why should we even call TSA when we know the request is invalid ?

Reply:
> Why should we even call TSA when we know the request is invalid ?

Make sence. Updating the code.

Reply:
Code updated.